src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -185,36 +188,64 @@ async Task EnsureLong()
                         // Or we consider it not timed out if we have already read everything (count == 0)
                         // In which case we reinitialize it on the next call to EnsureLong()
                         if (async)
-                            _timeoutCts.RestartTimeoutWithoutReset();
-
+                            Cts.RestartTimeoutWithoutReset();
                     }
-
-                    _timeoutCts.Stop();
-                    NpgsqlEventSource.Log.BytesRead(totalRead);
-                }
-                catch (Exception e)
-                {
-                    // Stopping twice (in case the previous Stop() call succeeded) doesn't hurt.
-                    // Not stopping will cause an assertion failure in debug mode when we call Start() the next time.
-                    // We can't stop in a finally block because Connector.Break() will dispose the buffer and the contained
-                    // _timeoutCts
-                    _timeoutCts.Stop();
-                    switch (e)
+                    catch (Exception e)
                     {
-                    // User requested the cancellation
-                    case OperationCanceledException _ when (cancellationToken.IsCancellationRequested):
-                        throw dontBreakOnCancellation ? e : Connector.Break(e);
-                    // Read timeout
-                    case OperationCanceledException _:
-                    // Note that mono throws SocketException with the wrong error (see #1330)
-                    case IOException _ when (e.InnerException as SocketException)?.SocketErrorCode ==
-                                            (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock):
-                        Debug.Assert(e is OperationCanceledException ? async : !async);
-                        throw new NpgsqlException("Exception while reading from stream", new TimeoutException("Timeout during reading attempt"));
+                        // Stopping twice (in case the previous Stop() call succeeded) doesn't hurt.
+                        // Not stopping will cause an assertion failure in debug mode when we call Start() the next time.
+                        // We can't stop in a finally block because Connector.Break() will dispose the buffer and the contained
+                        // _timeoutCts
+                        Cts.Stop();
+                        switch (e)
+                        {
+                        // User requested the cancellation
+                        case OperationCanceledException _ when (cancellationToken.IsCancellationRequested):
```

Comment:
So is this now only relevant when we're being called from WaitAsync? If so, then I think readingNotifications must be true (we can assert)